// @pebula/ngrid-material/__ivy_ngcc__/fesm2015/pebula-ngrid-material.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PblNgridCheckboxModule } from '@pebula/ngrid-material/selection-column';
import { PblNgridPaginatorModule } from '@pebula/ngrid-material/paginator';
import { PblNgridMatSortModule } from '@pebula/ngrid-material/sort';
import { PblNgridCellTooltipModule } from '@pebula/ngrid-material/cell-tooltip';
import { PblNgridContextMenuModule } from '@pebula/ngrid-material/context-menu';

/**
 * @fileoverview added by tsickle
 * Generated from: lib/ngrid-material.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as ɵngcc0 from '@angular/core';
class PblNgridMaterialModule {
}
PblNgridMaterialModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: PblNgridMaterialModule });
PblNgridMaterialModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function PblNgridMaterialModule_Factory(t) { return new (t || PblNgridMaterialModule)(); }, imports: [[
            CommonModule,
            PblNgridCheckboxModule,
            PblNgridPaginatorModule,
            PblNgridMatSortModule,
            PblNgridCellTooltipModule,
            PblNgridContextMenuModule,
        ],
        PblNgridCheckboxModule,
        PblNgridPaginatorModule,
        PblNgridMatSortModule,
        PblNgridCellTooltipModule,
        PblNgridContextMenuModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(PblNgridMaterialModule, { imports: function () { return [CommonModule,
        PblNgridCheckboxModule,
        PblNgridPaginatorModule,
        PblNgridMatSortModule,
        PblNgridCellTooltipModule,
        PblNgridContextMenuModule]; }, exports: function () { return [PblNgridCheckboxModule,
        PblNgridPaginatorModule,
        PblNgridMatSortModule,
        PblNgridCellTooltipModule,
        PblNgridContextMenuModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(PblNgridMaterialModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    PblNgridCheckboxModule,
                    PblNgridPaginatorModule,
                    PblNgridMatSortModule,
                    PblNgridCellTooltipModule,
                    PblNgridContextMenuModule,
                ],
                exports: [
                    PblNgridCheckboxModule,
                    PblNgridPaginatorModule,
                    PblNgridMatSortModule,
                    PblNgridCellTooltipModule,
                    PblNgridContextMenuModule,
                ]
            }]
    }], null, null); })();

/**
 * @fileoverview added by tsickle
 * Generated from: index.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: pebula-ngrid-material.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { PblNgridMaterialModule };
