// @pebula/ngrid-material/__ivy_ngcc__/fesm2015/pebula-ngrid-material-context-menu.js
import { Directive, Input, Component, ViewEncapsulation, ElementRef, ViewChild, ViewContainerRef, NgModule, Optional, SkipSelf, ComponentFactoryResolver } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MatIconModule } from '@angular/material/icon';
import { MatButtonModule } from '@angular/material/button';
import { MatMenu, MatMenuModule } from '@angular/material/menu';
import { MatFormFieldModule } from '@angular/material/form-field';
import { MatInputModule } from '@angular/material/input';
import { PblNgridPluginController, NgridPlugin, PblNgridMultiComponentRegistry, PblNgridModule, PblNgridRegistryService, PblNgridConfigService } from '@pebula/ngrid';
import { PblNgridOverlayPanelFactory, PblNgridOverlayPanelRef, PblNgridOverlayPanelComponentExtension, PblNgridOverlayPanelModule } from '@pebula/ngrid/overlay-panel';
import { __decorate, __metadata } from 'tslib';

/**
 * @fileoverview added by tsickle
 * Generated from: lib/header-context/header-context-menu.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@pebula/ngrid/overlay-panel';
import * as ɵngcc2 from '@pebula/ngrid';
import * as ɵngcc3 from '@angular/material/icon';
import * as ɵngcc4 from '@angular/material/menu';
import * as ɵngcc5 from '@angular/common';
import * as ɵngcc6 from '@angular/material/form-field';
import * as ɵngcc7 from '@angular/material/input';
import * as ɵngcc8 from '@angular/material/button';

const _c0 = ["mat-header-context-menu-trigger", ""];
const _c1 = ["columnMenu"];
const _c2 = ["menuViewLocation"];
function MatExcelStyleHeaderMenu_button_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "button", 3);
    ɵngcc0.ɵɵelementStart(1, "mat-icon");
    ɵngcc0.ɵɵtext(2, "sort");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(3, "Sort ");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵnextContext();
    const _r2 = ɵngcc0.ɵɵreference(16);
    ɵngcc0.ɵɵproperty("matMenuTriggerFor", _r2);
} }
const PLUGIN_KEY = 'matHeaderContextMenu';
let PblNgridMatHeaderContextMenuPlugin = class PblNgridMatHeaderContextMenuPlugin {
    /**
     * @param {?} overlayPanelFactory
     * @param {?} pluginCtrl
     */
    constructor(overlayPanelFactory, pluginCtrl) {
        this.pluginCtrl = pluginCtrl;
        this.overlayPanel = overlayPanelFactory.create(pluginCtrl.extApi.grid);
    }
};
PblNgridMatHeaderContextMenuPlugin.ɵfac = function PblNgridMatHeaderContextMenuPlugin_Factory(t) { return new (t || PblNgridMatHeaderContextMenuPlugin)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.PblNgridOverlayPanelFactory), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.PblNgridPluginController)); };
PblNgridMatHeaderContextMenuPlugin.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: PblNgridMatHeaderContextMenuPlugin, selectors: [["pbl-ngrid", "matHeaderContextMenu", ""]], inputs: { style: ["matHeaderContextMenu", "style"], config: "config" }, features: [ɵngcc0.ɵɵProvidersFeature([PblNgridOverlayPanelFactory])] });
/** @nocollapse */
PblNgridMatHeaderContextMenuPlugin.ctorParameters = () => [
    { type: PblNgridOverlayPanelFactory },
    { type: PblNgridPluginController }
];
PblNgridMatHeaderContextMenuPlugin.propDecorators = {
    style: [{ type: Input, args: ['matHeaderContextMenu',] }],
    config: [{ type: Input }]
};
PblNgridMatHeaderContextMenuPlugin = __decorate([
    NgridPlugin({ id: PLUGIN_KEY }),
    __metadata("design:paramtypes", [PblNgridOverlayPanelFactory,
        PblNgridPluginController])
], PblNgridMatHeaderContextMenuPlugin);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(PblNgridMatHeaderContextMenuPlugin, [{
        type: Directive,
        args: [{ selector: 'pbl-ngrid[matHeaderContextMenu]', providers: [PblNgridOverlayPanelFactory] }]
    }], function () { return [{ type: ɵngcc1.PblNgridOverlayPanelFactory }, { type: ɵngcc2.PblNgridPluginController }]; }, { style: [{
            type: Input,
            args: ['matHeaderContextMenu']
        }], config: [{
            type: Input
        }] }); })();
if (false) {
    /** @type {?} */
    PblNgridMatHeaderContextMenuPlugin.prototype.style;
    /** @type {?} */
    PblNgridMatHeaderContextMenuPlugin.prototype.config;
    /** @type {?} */
    PblNgridMatHeaderContextMenuPlugin.prototype.overlayPanel;
    /** @type {?} */
    PblNgridMatHeaderContextMenuPlugin.prototype.pluginCtrl;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/header-context/header-context-menu-trigger.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const DEFAULT_CONFIG = { hasBackdrop: true, xPos: 'after', yPos: 'below' };
class MatHeaderContextMenuTrigger {
    /**
     * @param {?} plugin
     * @param {?} elRef
     */
    constructor(plugin, elRef) {
        this.plugin = plugin;
        this.elRef = elRef;
    }
    /**
     * @return {?}
     */
    openOverlayPanel() {
        /** @type {?} */
        const config = this.plugin.config || DEFAULT_CONFIG;
        this.plugin.overlayPanel.open(this.plugin.style, this.elRef, config, this.context);
    }
}
MatHeaderContextMenuTrigger.ɵfac = function MatHeaderContextMenuTrigger_Factory(t) { return new (t || MatHeaderContextMenuTrigger)(ɵngcc0.ɵɵdirectiveInject(PblNgridMatHeaderContextMenuPlugin), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
MatHeaderContextMenuTrigger.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: MatHeaderContextMenuTrigger, selectors: [["div", "mat-header-context-menu-trigger", ""]], hostAttrs: [1, "mat-header-context-menu-trigger"], hostBindings: function MatHeaderContextMenuTrigger_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function MatHeaderContextMenuTrigger_click_HostBindingHandler() { return ctx.openOverlayPanel(); });
    } }, attrs: _c0, decls: 2, vars: 0, consts: [[2, "height", "16px", "width", "16px", "font-size", "16px", "line-height", "16px"]], template: function MatHeaderContextMenuTrigger_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "mat-icon", 0);
        ɵngcc0.ɵɵtext(1, "more_vert");
        ɵngcc0.ɵɵelementEnd();
    } }, directives: [ɵngcc3.MatIcon], styles: ["div.mat-header-context-menu-trigger{position:absolute;display:flex;align-items:center;right:0;height:100%;cursor:pointer;margin-right:12px;z-index:100}"], encapsulation: 2 });
/** @nocollapse */
MatHeaderContextMenuTrigger.ctorParameters = () => [
    { type: PblNgridMatHeaderContextMenuPlugin },
    { type: ElementRef }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MatHeaderContextMenuTrigger, [{
        type: Component,
        args: [{
                selector: 'div[mat-header-context-menu-trigger]',
                host: {
                    class: 'mat-header-context-menu-trigger',
                    '(click)': 'openOverlayPanel()'
                },
                template: "<mat-icon style=\"height: 16px; width: 16px; font-size: 16px; line-height: 16px;\">more_vert</mat-icon>\n",
                encapsulation: ViewEncapsulation.None,
                styles: ["div.mat-header-context-menu-trigger{position:absolute;display:flex;align-items:center;right:0;height:100%;cursor:pointer;margin-right:12px;z-index:100}"]
            }]
    }], function () { return [{ type: PblNgridMatHeaderContextMenuPlugin }, { type: ɵngcc0.ElementRef }]; }, null); })();
if (false) {
    /** @type {?} */
    MatHeaderContextMenuTrigger.prototype.context;
    /**
     * @type {?}
     * @private
     */
    MatHeaderContextMenuTrigger.prototype.plugin;
    /**
     * @type {?}
     * @private
     */
    MatHeaderContextMenuTrigger.prototype.elRef;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/header-context/header-context-menu-extension.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class MatHeaderContextMenuExtension extends PblNgridMultiComponentRegistry {
    /**
     * @param {?} cfr
     */
    constructor(cfr) {
        super();
        this.cfr = cfr;
        this.name = 'matHeaderContextMenuTrigger';
        this.kind = 'dataHeaderExtensions';
        this.projectContent = false;
    }
    /**
     * @param {?} context
     * @return {?}
     */
    shouldRender(context) {
        return !!context.injector.get(PblNgridMatHeaderContextMenuPlugin, false);
    }
    /**
     * @param {?} context
     * @return {?}
     */
    getFactory(context) {
        return this.cfr.resolveComponentFactory(MatHeaderContextMenuTrigger);
    }
    /**
     * @param {?} context
     * @param {?} cmpRef
     * @return {?}
     */
    onCreated(context, cmpRef) {
        cmpRef.instance.context = context;
        cmpRef.changeDetectorRef.markForCheck();
    }
}
if (false) {
    /** @type {?} */
    MatHeaderContextMenuExtension.prototype.name;
    /** @type {?} */
    MatHeaderContextMenuExtension.prototype.kind;
    /** @type {?} */
    MatHeaderContextMenuExtension.prototype.projectContent;
    /**
     * @type {?}
     * @private
     */
    MatHeaderContextMenuExtension.prototype.cfr;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/header-context/styles/excel-style-header-menu.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class MatExcelStyleHeaderMenu {
    /**
     * @param {?} ref
     */
    constructor(ref) {
        this.ref = ref;
        this.currentFilter = '';
        this.column = ref.data.col;
        this.grid = ref.data.grid;
        if (this.grid.ds.sort.column === this.column) {
            this.currentSort = this.grid.ds.sort.sort.order;
        }
        this.currentPin = this.column.columnDef.sticky ? 'start' : this.column.columnDef.stickyEnd ? 'end' : undefined;
        /** @type {?} */
        const dsFilter = this.grid.ds.filter;
        if (dsFilter && dsFilter.type === 'value' && dsFilter.columns && dsFilter.columns.indexOf(this.column) >= 0) {
            this.currentFilter = dsFilter.filter;
        }
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.matMenu.closed.subscribe((/**
         * @param {?} reason
         * @return {?}
         */
        reason => {
            this.ref.close();
        }));
        /** @type {?} */
        const view = this.menuViewLocation.createEmbeddedView(this.matMenu.templateRef);
        this.matMenu.setElevation(0);
        this.matMenu.focusFirstItem('program');
        this.matMenu._resetAnimation();
        view.markForCheck();
        view.detectChanges();
        this.matMenu._startAnimation();
    }
    /**
     * @return {?}
     */
    hide() {
        /** @type {?} */
        const hidden = [this.column.id];
        for (const col of this.grid.columnApi.columns) {
            if (col.hidden) {
                hidden.push(col.id);
            }
        }
        this.grid.hideColumns = hidden;
    }
    /**
     * @param {?} sort
     * @return {?}
     */
    onSortToggle(sort) {
        if (this.currentSort === sort) {
            this.grid.ds.setSort();
        }
        else {
            this.grid.ds.setSort(this.column, { order: sort });
        }
    }
    /**
     * @param {?} pin
     * @return {?}
     */
    onPinToggle(pin) {
        if (this.currentPin === pin) {
            this.column.columnDef.updatePin();
        }
        else {
            this.column.columnDef.updatePin(pin);
        }
    }
    /**
     * @param {?} filterValue
     * @return {?}
     */
    filterColumn(filterValue) {
        this.currentFilter = filterValue;
        if (!filterValue) {
            this.grid.setFilter();
        }
        else {
            this.grid.setFilter(filterValue.trim(), [this.column]);
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    clickTrap(event) {
        event.preventDefault();
        event.stopPropagation();
    }
}
MatExcelStyleHeaderMenu.ɵfac = function MatExcelStyleHeaderMenu_Factory(t) { return new (t || MatExcelStyleHeaderMenu)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.PblNgridOverlayPanelRef)); };
MatExcelStyleHeaderMenu.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: MatExcelStyleHeaderMenu, selectors: [["mat-excel-style-header-menu"]], viewQuery: function MatExcelStyleHeaderMenu_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c1, true, MatMenu);
        ɵngcc0.ɵɵstaticViewQuery(_c2, true, ViewContainerRef);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.matMenu = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.menuViewLocation = _t.first);
    } }, decls: 47, vars: 15, consts: [[1, "pbl-mat-menu-panel"], ["columnMenu", "matMenu"], ["mat-menu-item", "", 3, "matMenuTriggerFor", 4, "ngIf"], ["mat-menu-item", "", 3, "matMenuTriggerFor"], ["mat-menu-item", "", 3, "click"], ["sortMenu", "matMenu"], [3, "color"], ["pinMenu", "matMenu"], [1, "mat-menu-item", "pbl-mat-menu-row", 3, "click"], ["matInput", "", 3, "value", "keyup"], ["matPrefix", ""], ["mat-button", "", "matSuffix", "", "mat-icon-button", "", "aria-label", "Clear", 3, "click"], ["menuViewLocation", ""]], template: function MatExcelStyleHeaderMenu_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "mat-menu", 0, 1);
        ɵngcc0.ɵɵtemplate(2, MatExcelStyleHeaderMenu_button_2_Template, 4, 1, "button", 2);
        ɵngcc0.ɵɵelementStart(3, "button", 3);
        ɵngcc0.ɵɵelementStart(4, "mat-icon");
        ɵngcc0.ɵɵtext(5, "place");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtext(6, "Pin ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(7, "button", 4);
        ɵngcc0.ɵɵlistener("click", function MatExcelStyleHeaderMenu_Template_button_click_7_listener() { return ctx.grid.columnApi.autoSizeColumn(ctx.column); });
        ɵngcc0.ɵɵelementStart(8, "mat-icon");
        ɵngcc0.ɵɵtext(9, "keyboard_tab");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtext(10, "Auto Fit ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "button", 4);
        ɵngcc0.ɵɵlistener("click", function MatExcelStyleHeaderMenu_Template_button_click_11_listener() { return ctx.hide(); });
        ɵngcc0.ɵɵelementStart(12, "mat-icon");
        ɵngcc0.ɵɵtext(13, "visibility_off");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtext(14, "Hide Column ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(15, "mat-menu", null, 5);
        ɵngcc0.ɵɵelementStart(17, "button", 4);
        ɵngcc0.ɵɵlistener("click", function MatExcelStyleHeaderMenu_Template_button_click_17_listener() { return ctx.onSortToggle("asc"); });
        ɵngcc0.ɵɵelementStart(18, "mat-icon", 6);
        ɵngcc0.ɵɵtext(19, "arrow_upward");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(20, "span");
        ɵngcc0.ɵɵtext(21, "Ascending");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(22, "button", 4);
        ɵngcc0.ɵɵlistener("click", function MatExcelStyleHeaderMenu_Template_button_click_22_listener() { return ctx.onSortToggle("desc"); });
        ɵngcc0.ɵɵelementStart(23, "mat-icon", 6);
        ɵngcc0.ɵɵtext(24, "arrow_downward");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(25, "span");
        ɵngcc0.ɵɵtext(26, "Descending");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(27, "mat-menu", null, 7);
        ɵngcc0.ɵɵelementStart(29, "button", 4);
        ɵngcc0.ɵɵlistener("click", function MatExcelStyleHeaderMenu_Template_button_click_29_listener() { return ctx.onPinToggle("start"); });
        ɵngcc0.ɵɵelementStart(30, "span");
        ɵngcc0.ɵɵtext(31, "Start");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(32, "button", 4);
        ɵngcc0.ɵɵlistener("click", function MatExcelStyleHeaderMenu_Template_button_click_32_listener() { return ctx.onPinToggle("end"); });
        ɵngcc0.ɵɵelementStart(33, "span");
        ɵngcc0.ɵɵtext(34, "End");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(35, "div", 8);
        ɵngcc0.ɵɵlistener("click", function MatExcelStyleHeaderMenu_Template_div_click_35_listener($event) { return ctx.clickTrap($event); });
        ɵngcc0.ɵɵelementStart(36, "mat-form-field");
        ɵngcc0.ɵɵelementStart(37, "mat-label");
        ɵngcc0.ɵɵtext(38, "Search");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(39, "input", 9);
        ɵngcc0.ɵɵlistener("keyup", function MatExcelStyleHeaderMenu_Template_input_keyup_39_listener($event) { return ctx.filterColumn($event.target.value); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(40, "mat-icon", 10);
        ɵngcc0.ɵɵtext(41, "search");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(42, "button", 11);
        ɵngcc0.ɵɵlistener("click", function MatExcelStyleHeaderMenu_Template_button_click_42_listener() { return ctx.filterColumn(""); });
        ɵngcc0.ɵɵelementStart(43, "mat-icon");
        ɵngcc0.ɵɵtext(44, "close");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementContainer(45, null, 12);
    } if (rf & 2) {
        const _r3 = ɵngcc0.ɵɵreference(28);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.column.sort);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("matMenuTriggerFor", _r3);
        ɵngcc0.ɵɵadvance(14);
        ɵngcc0.ɵɵclassProp("menu-item-selected", ctx.currentSort === "asc");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("color", ctx.currentSort === "asc" ? "primary" : "");
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵclassProp("menu-item-selected", ctx.currentSort === "desc");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("color", ctx.currentSort === "desc" ? "primary" : "");
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵclassProp("menu-item-selected", ctx.currentPin === "start");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵclassProp("menu-item-selected", ctx.currentPin === "end");
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵproperty("value", ctx.currentFilter);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵstyleProp("visibility", ctx.currentFilter ? "visible" : "hidden");
    } }, directives: [ɵngcc4._MatMenu, ɵngcc5.NgIf, ɵngcc4.MatMenuItem, ɵngcc4.MatMenuTrigger, ɵngcc3.MatIcon, ɵngcc6.MatFormField, ɵngcc6.MatLabel, ɵngcc7.MatInput, ɵngcc6.MatPrefix, ɵngcc8.MatButton, ɵngcc6.MatSuffix], styles: [".mat-menu-panel.pbl-mat-menu-panel{max-width:400px}.mat-menu-item.pbl-mat-menu-row{width:100%;box-sizing:border-box;line-height:inherit;height:auto;margin:6px 0;cursor:inherit}.mat-menu-item.pbl-mat-menu-row:hover{background:inherit}"], encapsulation: 2 });
/** @nocollapse */
MatExcelStyleHeaderMenu.ctorParameters = () => [
    { type: PblNgridOverlayPanelRef }
];
MatExcelStyleHeaderMenu.propDecorators = {
    matMenu: [{ type: ViewChild, args: ['columnMenu', { read: MatMenu, static: true },] }],
    menuViewLocation: [{ type: ViewChild, args: ['menuViewLocation', { read: ViewContainerRef, static: true },] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MatExcelStyleHeaderMenu, [{
        type: Component,
        args: [{
                selector: 'mat-excel-style-header-menu',
                template: "<mat-menu #columnMenu=\"matMenu\" class=\"pbl-mat-menu-panel\">\n\n  <button *ngIf=\"column.sort\" mat-menu-item [matMenuTriggerFor]=\"sortMenu\">\n    <mat-icon>sort</mat-icon>Sort\n  </button>\n  <button mat-menu-item [matMenuTriggerFor]=\"pinMenu\">\n    <mat-icon>place</mat-icon>Pin\n  </button>\n  <button mat-menu-item (click)=\"grid.columnApi.autoSizeColumn(column)\">\n    <mat-icon>keyboard_tab</mat-icon>Auto Fit\n  </button>\n  <button mat-menu-item (click)=\"hide()\">\n    <mat-icon>visibility_off</mat-icon>Hide Column\n  </button>\n\n  <mat-menu #sortMenu=\"matMenu\">\n    <button mat-menu-item (click)=\"onSortToggle('asc')\" [class.menu-item-selected]=\"currentSort === 'asc'\">\n      <mat-icon [color]=\"currentSort === 'asc' ? 'primary' : ''\">arrow_upward</mat-icon>\n      <span>Ascending</span>\n    </button>\n    <button mat-menu-item (click)=\"onSortToggle('desc')\" [class.menu-item-selected]=\"currentSort === 'desc'\">\n      <mat-icon [color]=\"currentSort === 'desc' ? 'primary' : ''\">arrow_downward</mat-icon>\n      <span>Descending</span>\n    </button>\n  </mat-menu>\n\n  <mat-menu #pinMenu=\"matMenu\">\n    <button mat-menu-item (click)=\"onPinToggle('start')\" [class.menu-item-selected]=\"currentPin === 'start'\">\n      <span>Start</span>\n    </button>\n    <button mat-menu-item (click)=\"onPinToggle('end')\" [class.menu-item-selected]=\"currentPin === 'end'\">\n      <span>End</span>\n    </button>\n  </mat-menu>\n\n  <div class=\"mat-menu-item pbl-mat-menu-row\" (click)=\"clickTrap($event)\">\n    <mat-form-field>\n      <mat-label>Search</mat-label>\n      <input matInput (keyup)=\"filterColumn($event.target.value)\" [value]=\"currentFilter\">\n      <mat-icon matPrefix>search</mat-icon>\n      <button mat-button [style.visibility]=\"currentFilter ? 'visible' : 'hidden'\" matSuffix mat-icon-button aria-label=\"Clear\"\n              (click)=\"filterColumn('')\">\n        <mat-icon>close</mat-icon>\n      </button>\n    </mat-form-field>\n  </div>\n</mat-menu>\n<ng-container #menuViewLocation></ng-container>\n",
                encapsulation: ViewEncapsulation.None,
                styles: [".mat-menu-panel.pbl-mat-menu-panel{max-width:400px}.mat-menu-item.pbl-mat-menu-row{width:100%;box-sizing:border-box;line-height:inherit;height:auto;margin:6px 0;cursor:inherit}.mat-menu-item.pbl-mat-menu-row:hover{background:inherit}"]
            }]
    }], function () { return [{ type: ɵngcc1.PblNgridOverlayPanelRef }]; }, { matMenu: [{
            type: ViewChild,
            args: ['columnMenu', { read: MatMenu, static: true }]
        }], menuViewLocation: [{
            type: ViewChild,
            args: ['menuViewLocation', { read: ViewContainerRef, static: true }]
        }] }); })();
if (false) {
    /** @type {?} */
    MatExcelStyleHeaderMenu.prototype.column;
    /** @type {?} */
    MatExcelStyleHeaderMenu.prototype.grid;
    /** @type {?} */
    MatExcelStyleHeaderMenu.prototype.matMenu;
    /** @type {?} */
    MatExcelStyleHeaderMenu.prototype.menuViewLocation;
    /** @type {?} */
    MatExcelStyleHeaderMenu.prototype.currentSort;
    /** @type {?} */
    MatExcelStyleHeaderMenu.prototype.currentPin;
    /** @type {?} */
    MatExcelStyleHeaderMenu.prototype.currentFilter;
    /**
     * @type {?}
     * @private
     */
    MatExcelStyleHeaderMenu.prototype.ref;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/context-menu.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class PblNgridContextMenuModule {
    /**
     * @param {?} parentModule
     * @param {?} registry
     * @param {?} cfr
     * @param {?} configService
     */
    constructor(parentModule, registry, cfr, configService) {
        if (parentModule) {
            return;
        }
        registry.addMulti('dataHeaderExtensions', new MatHeaderContextMenuExtension(cfr));
        registry.addMulti('overlayPanels', new PblNgridOverlayPanelComponentExtension('excelMenu', MatExcelStyleHeaderMenu, cfr));
    }
}
PblNgridContextMenuModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: PblNgridContextMenuModule });
PblNgridContextMenuModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function PblNgridContextMenuModule_Factory(t) { return new (t || PblNgridContextMenuModule)(ɵngcc0.ɵɵinject(PblNgridContextMenuModule, 12), ɵngcc0.ɵɵinject(ɵngcc2.PblNgridRegistryService), ɵngcc0.ɵɵinject(ɵngcc0.ComponentFactoryResolver), ɵngcc0.ɵɵinject(ɵngcc2.PblNgridConfigService)); }, imports: [[
            CommonModule,
            MatIconModule,
            MatButtonModule,
            MatMenuModule,
            MatFormFieldModule,
            MatInputModule,
            PblNgridModule,
            PblNgridOverlayPanelModule,
        ]] });
/** @nocollapse */
PblNgridContextMenuModule.ctorParameters = () => [
    { type: PblNgridContextMenuModule, decorators: [{ type: Optional }, { type: SkipSelf }] },
    { type: PblNgridRegistryService },
    { type: ComponentFactoryResolver },
    { type: PblNgridConfigService }
];
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(PblNgridContextMenuModule, { declarations: function () { return [MatHeaderContextMenuTrigger,
        PblNgridMatHeaderContextMenuPlugin,
        MatExcelStyleHeaderMenu]; }, imports: function () { return [CommonModule,
        MatIconModule,
        MatButtonModule,
        MatMenuModule,
        MatFormFieldModule,
        MatInputModule,
        PblNgridModule,
        PblNgridOverlayPanelModule]; }, exports: function () { return [PblNgridMatHeaderContextMenuPlugin]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(PblNgridContextMenuModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    MatIconModule,
                    MatButtonModule,
                    MatMenuModule,
                    MatFormFieldModule,
                    MatInputModule,
                    PblNgridModule,
                    PblNgridOverlayPanelModule,
                ],
                declarations: [
                    MatHeaderContextMenuTrigger,
                    PblNgridMatHeaderContextMenuPlugin,
                    MatExcelStyleHeaderMenu,
                ],
                exports: [
                    PblNgridMatHeaderContextMenuPlugin,
                ],
                entryComponents: [
                    // TODO: remove when ViewEngine is no longer supported by angular (V11 ???)
                    MatHeaderContextMenuTrigger,
                    MatExcelStyleHeaderMenu,
                ]
            }]
    }], function () { return [{ type: PblNgridContextMenuModule, decorators: [{
                type: Optional
            }, {
                type: SkipSelf
            }] }, { type: ɵngcc2.PblNgridRegistryService }, { type: ɵngcc0.ComponentFactoryResolver }, { type: ɵngcc2.PblNgridConfigService }]; }, null); })();

/**
 * @fileoverview added by tsickle
 * Generated from: index.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: pebula-ngrid-material-context-menu.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { PblNgridContextMenuModule, MatHeaderContextMenuTrigger as ɵa, PblNgridMatHeaderContextMenuPlugin as ɵb, MatExcelStyleHeaderMenu as ɵc };
