// @pebula/ngrid/esm2015/lib/grid/context/api.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/grid/context/api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BehaviorSubject, Subject, asapScheduler } from 'rxjs';
import { debounceTime, buffer, map, filter } from 'rxjs/operators';
import { removeFromArray } from '@pebula/utils';
import { findRowRenderedIndex, resolveCellReference } from './utils';
import { PblRowContext } from './row';
import { PblCellContext } from './cell';
/**
 * @template T
 */
export class ContextApi {
    /**
     * @param {?} extApi
     */
    constructor(extApi) {
        this.extApi = extApi;
        this.viewCache = new Map();
        this.cache = new Map();
        this.activeSelected = [];
        this.focusChanged$ = new BehaviorSubject({ prev: undefined, curr: undefined });
        this.selectionChanged$ = new Subject();
        /**
         * Notify when the focus has changed.
         *
         * > Note that the notification is not immediate, it will occur on the closest micro-task after the change.
         */
        this.focusChanged = this.focusChanged$
            .pipe(buffer(this.focusChanged$.pipe(debounceTime(0, asapScheduler))), map((/**
         * @param {?} events
         * @return {?}
         */
        events => ({ prev: events[0].prev, curr: events[events.length - 1].curr }))));
        /**
         * Notify when the selected cells has changed.
         */
        this.selectionChanged = this.selectionChanged$.asObservable();
        this.vcRef = extApi.cdkTable._rowOutlet.viewContainer;
        this.columnApi = extApi.grid.columnApi;
        extApi.events
            .pipe(filter((/**
         * @param {?} e
         * @return {?}
         */
        e => e.kind === 'onDestroy')))
            .subscribe((/**
         * @param {?} e
         * @return {?}
         */
        e => this.destroy()));
        /** @type {?} */
        const updateContext = (/**
         * @return {?}
         */
        () => {
            /** @type {?} */
            const viewPortRect = this.getViewRect();
            /** @type {?} */
            const lastView = new Set(Array.from(this.viewCache.values()).map((/**
             * @param {?} v
             * @return {?}
             */
            v => v.identity)));
            /** @type {?} */
            const unmatchedRefs = new Map();
            /** @type {?} */
            let keepProcessOutOfView = !!viewPortRect;
            for (let i = 0, len = this.vcRef.length; i < len; i++) {
                /** @type {?} */
                const viewRef = this.findViewRef(i);
                /** @type {?} */
                const rowContext = this.findRowContext(viewRef, i);
                this.viewCache.set(i, rowContext);
                lastView.delete(rowContext.identity);
                // Identity did not change but context did change
                // This is probably due to trackBy with index reference or that matched data on some property but the actual data reference changed.
                // We log these and handle them later, they come in pair and we need to switch the context between the values in the pair.
                // The pair is a 2 item tuple - 1st item is new index, 2nd item is the old index.
                // We build the pairs, each pair is a switch
                if (viewRef.context.$implicit !== rowContext.$implicit) {
                    /** @type {?} */
                    let pair = unmatchedRefs.get(rowContext.$implicit) || [-1, -1];
                    pair[1] = i;
                    unmatchedRefs.set(rowContext.$implicit, pair);
                    pair = unmatchedRefs.get(viewRef.context.$implicit) || [-1, -1];
                    pair[0] = i;
                    unmatchedRefs.set(viewRef.context.$implicit, pair);
                }
                if (keepProcessOutOfView) {
                    keepProcessOutOfView = processOutOfView(viewRef, viewPortRect, 'top');
                }
            }
            if (unmatchedRefs.size > 0) {
                // We have pairs but we can't just start switching because when the items move or swap we need
                // to update their values and so we need to cache one of them.
                // The operation will effect all items (N) between then origin and destination.
                // When N === 2 its a swap, when N > 2 its a move.
                // In both cases the first and last operations share the same object.
                // Also, we need to make sure that the order of operations does not use the same row as the source more then once.
                // For example, If I copy row 5 to to row 4 and then 4 to 3 I need to start from 3->4->5, if I do 5->4->3 I will get 5 in all rows.
                //
                // We use the source (pair[1]) for sorting, the sort order depends on the direction of the move (up/down).
                /** @type {?} */
                const arr = Array.from(unmatchedRefs.entries()).filter((/**
                 * @param {?} entry
                 * @return {?}
                 */
                entry => {
                    /** @type {?} */
                    const pair = entry[1];
                    if (pair[0] === -1) {
                        return false;
                    }
                    else if (pair[1] === -1) {
                        /** @type {?} */
                        const to = this.viewCache.get(pair[0]);
                        to.$implicit = entry[0];
                        return false;
                    }
                    return true;
                })).map((/**
                 * @param {?} entry
                 * @return {?}
                 */
                entry => entry[1]));
                unmatchedRefs.clear();
                if (arr.length) {
                    /** @type {?} */
                    const sortFn = arr[arr.length - 1][0] - arr[arr.length - 1][1] > 0 // check sort direction
                        ? (/**
                         * @param {?} a
                         * @param {?} b
                         * @return {?}
                         */
                        (a, b) => b[1] - a[1])
                        : (/**
                         * @param {?} a
                         * @param {?} b
                         * @return {?}
                         */
                        (a, b) => a[1] - b[1]);
                    arr.sort(sortFn);
                    /** @type {?} */
                    const lastOp = {
                        data: this.viewCache.get(arr[0][0]).$implicit,
                        state: this.viewCache.get(arr[0][0]).getState(),
                        pair: arr.pop(),
                    };
                    for (const pair of arr) {
                        // What we're doing here is switching the context wrapped by `RotContext` while the `RowContext` preserve it's identity.
                        // Each row context has a state, which is valid for it's current context, if we switch context we must switch state as well and also
                        // cache it.
                        /** @type {?} */
                        const to = this.viewCache.get(pair[0]);
                        /** @type {?} */
                        const from = this.viewCache.get(pair[1]);
                        /** @type {?} */
                        const state = from.getState();
                        state.identity = to.identity;
                        this.cache.set(to.identity, state);
                        to.fromState(state);
                        to.$implicit = from.$implicit;
                    }
                    /** @type {?} */
                    const to = this.viewCache.get(lastOp.pair[0]);
                    lastOp.state.identity = to.identity;
                    this.cache.set(to.identity, lastOp.state);
                    to.fromState(lastOp.state);
                    to.$implicit = lastOp.data;
                }
            }
            if (viewPortRect) {
                for (let i = this.vcRef.length - 1; i > -1; i--) {
                    if (!processOutOfView(this.findViewRef(i), viewPortRect, 'bottom')) {
                        break;
                    }
                }
            }
            lastView.forEach((/**
             * @param {?} ident
             * @return {?}
             */
            ident => this.cache.get(ident).firstRender = false));
        });
        updateContext();
        extApi.cdkTable.onRenderRows.subscribe(updateContext);
    }
    /**
     * The reference to currently focused cell context.
     * You can retrieve the actual context or context cell using `findRowInView` and / or `findRowInCache`.
     *
     * > Note that when virtual scroll is enabled the currently focused cell does not have to exist in the view.
     * If this is the case `findRowInView` will return undefined, use `findRowInCache` instead.
     * @return {?}
     */
    get focusedCell() {
        return this.activeFocused ? Object.assign({}, this.activeFocused) : undefined;
    }
    /**
     * The reference to currently selected range of cell's context.
     * You can retrieve the actual context or context cell using `findRowInView` and / or `findRowInCache`.
     *
     * > Note that when virtual scroll is enabled the currently selected cells does not have to exist in the view.
     * If this is the case `findRowInView` will return undefined, use `findRowInCache` instead.
     * @return {?}
     */
    get selectedCells() {
        return this.activeSelected.slice();
    }
    /**
     * Focus the provided cell.
     * If a cell is not provided will un-focus (blur) the currently focused cell (if there is one).
     * @param {?=} cellRef A Reference to the cell
     * @param {?=} markForCheck Mark the row for change detection
     * @return {?}
     */
    focusCell(cellRef, markForCheck) {
        if (!cellRef || cellRef === true) {
            if (this.activeFocused) {
                const { rowIdent, colIndex } = this.activeFocused;
                this.activeFocused = undefined;
                this.updateState(rowIdent, colIndex, { focused: false });
                this.emitFocusChanged(this.activeFocused);
                if (markForCheck) {
                    /** @type {?} */
                    const rowContext = this.findRowInView(rowIdent);
                    if (rowContext) {
                        this.extApi.grid._cdkTable.syncRows('data', rowContext.index);
                    }
                }
            }
        }
        else {
            /** @type {?} */
            const ref = resolveCellReference(cellRef, (/** @type {?} */ (this)));
            if (ref) {
                this.focusCell(markForCheck);
                if (ref instanceof PblCellContext) {
                    if (!ref.focused && !this.extApi.grid.viewport.isScrolling) {
                        this.updateState(ref.rowContext.identity, ref.index, { focused: true });
                        this.activeFocused = { rowIdent: ref.rowContext.identity, colIndex: ref.index };
                        this.selectCells([this.activeFocused], markForCheck, true);
                        if (markForCheck) {
                            this.extApi.grid._cdkTable.syncRows('data', ref.rowContext.index);
                        }
                    }
                }
                else {
                    this.updateState(ref[0].identity, ref[1], { focused: true });
                    this.activeFocused = { rowIdent: ref[0].identity, colIndex: ref[1] };
                }
                this.emitFocusChanged(this.activeFocused);
            }
        }
    }
    /**
     * Select all provided cells.
     * @param {?} cellRefs
     * @param {?=} markForCheck Mark the row for change detection
     * @param {?=} clearCurrent Clear the current selection before applying the new selection.
     * Default to false (add to current).
     * @return {?}
     */
    selectCells(cellRefs, markForCheck, clearCurrent) {
        /** @type {?} */
        const toMarkRendered = new Set();
        if (clearCurrent) {
            this.unselectCells();
        }
        /** @type {?} */
        const added = [];
        for (const cellRef of cellRefs) {
            /** @type {?} */
            const ref = resolveCellReference(cellRef, (/** @type {?} */ (this)));
            if (ref instanceof PblCellContext) {
                if (!ref.selected && !this.extApi.grid.viewport.isScrolling) {
                    /** @type {?} */
                    const rowIdent = ref.rowContext.identity;
                    /** @type {?} */
                    const colIndex = ref.index;
                    this.updateState(rowIdent, colIndex, { selected: true });
                    /** @type {?} */
                    const dataPoint = { rowIdent, colIndex };
                    this.activeSelected.push(dataPoint);
                    added.push(dataPoint);
                    if (markForCheck) {
                        toMarkRendered.add(ref.rowContext.index);
                    }
                }
            }
            else if (ref) {
                const [rowState, colIndex] = ref;
                if (!rowState.cells[colIndex].selected) {
                    this.updateState(rowState.identity, colIndex, { selected: true });
                    this.activeSelected.push({ rowIdent: rowState.identity, colIndex });
                }
            }
        }
        if (toMarkRendered.size > 0) {
            this.extApi.grid._cdkTable.syncRows('data', ...Array.from(toMarkRendered.values()));
        }
        this.selectionChanged$.next({ added, removed: [] });
    }
    /**
     * Unselect all provided cells.
     * If cells are not provided will un-select all currently selected cells.
     * @param {?=} cellRefs
     * @param {?=} markForCheck Mark the row for change detection
     * @return {?}
     */
    unselectCells(cellRefs, markForCheck) {
        /** @type {?} */
        const toMarkRendered = new Set();
        /** @type {?} */
        let toUnselect = this.activeSelected;
        /** @type {?} */
        let removeAll = true;
        if (Array.isArray(cellRefs)) {
            toUnselect = cellRefs;
            removeAll = false;
        }
        else {
            markForCheck = !!cellRefs;
            this.activeSelected = [];
        }
        /** @type {?} */
        const removed = [];
        for (const cellRef of toUnselect) {
            /** @type {?} */
            const ref = resolveCellReference(cellRef, (/** @type {?} */ (this)));
            if (ref instanceof PblCellContext) {
                if (ref.selected) {
                    /** @type {?} */
                    const rowIdent = ref.rowContext.identity;
                    /** @type {?} */
                    const colIndex = ref.index;
                    this.updateState(rowIdent, colIndex, { selected: false });
                    if (!removeAll) {
                        /** @type {?} */
                        const wasRemoved = removeFromArray(this.activeSelected, (/**
                         * @param {?} item
                         * @return {?}
                         */
                        item => item.colIndex === colIndex && item.rowIdent === rowIdent));
                        if (wasRemoved) {
                            removed.push({ rowIdent, colIndex });
                        }
                    }
                    if (markForCheck) {
                        toMarkRendered.add(ref.rowContext.index);
                    }
                }
            }
            else if (ref) {
                const [rowState, colIndex] = ref;
                if (rowState.cells[colIndex].selected) {
                    this.updateState(rowState.identity, colIndex, { selected: false });
                    if (!removeAll) {
                        /** @type {?} */
                        const wasRemoved = removeFromArray(this.activeSelected, (/**
                         * @param {?} item
                         * @return {?}
                         */
                        item => item.colIndex === colIndex && item.rowIdent === rowState.identity));
                        if (wasRemoved) {
                            removed.push({ rowIdent: rowState.identity, colIndex });
                        }
                    }
                }
            }
        }
        if (toMarkRendered.size > 0) {
            this.extApi.grid._cdkTable.syncRows('data', ...Array.from(toMarkRendered.values()));
        }
        this.selectionChanged$.next({ added: [], removed });
    }
    /**
     * @return {?}
     */
    clear() {
        for (let i = 0, len = this.vcRef.length; i < len; i++) {
            /** @type {?} */
            const viewRef = this.findViewRef(i);
            viewRef.context.pblRowContext = undefined;
        }
        this.viewCache.clear();
        this.cache.clear();
    }
    /**
     * @param {?} row
     * @return {?}
     */
    getRow(row) {
        /** @type {?} */
        const index = typeof row === 'number' ? row : findRowRenderedIndex(row);
        return this.rowContext(index);
    }
    /**
     * @param {?} rowOrCellElement
     * @param {?=} col
     * @return {?}
     */
    getCell(rowOrCellElement, col) {
        if (typeof rowOrCellElement === 'number') {
            /** @type {?} */
            const rowContext = this.rowContext(rowOrCellElement);
            if (rowContext) {
                return rowContext.cell(col);
            }
        }
        else {
            /** @type {?} */
            const ref = resolveCellReference(rowOrCellElement, (/** @type {?} */ (this)));
            if (ref instanceof PblCellContext) {
                return ref;
            }
        }
    }
    /**
     * @param {?} cell
     * @return {?}
     */
    getDataItem(cell) {
        /** @type {?} */
        const ref = resolveCellReference(cell, (/** @type {?} */ (this)));
        if (ref instanceof PblCellContext) {
            return ref.col.getValue(ref.rowContext.$implicit);
        }
        else if (ref) {
            /** @type {?} */
            const row = this.extApi.grid.ds.source[ref[0].dataIndex];
            /** @type {?} */
            const column = this.extApi.grid.columnApi.findColumnAt(ref[1]);
            return column.getValue(row);
        }
    }
    /**
     * @param {?} renderRowIndex
     * @param {?} column
     * @return {?}
     */
    createCellContext(renderRowIndex, column) {
        /** @type {?} */
        const rowContext = this.rowContext(renderRowIndex);
        /** @type {?} */
        const colIndex = this.columnApi.indexOf(column);
        return rowContext.cell(colIndex);
    }
    /**
     * @param {?} renderRowIndex
     * @return {?}
     */
    rowContext(renderRowIndex) {
        return this.viewCache.get(renderRowIndex);
    }
    /**
     * @param {?} rowContext
     * @return {?}
     */
    updateOutOfViewState(rowContext) {
        /** @type {?} */
        const viewPortRect = this.getViewRect();
        /** @type {?} */
        const viewRef = this.findViewRef(rowContext.index);
        processOutOfView(viewRef, viewPortRect);
    }
    /**
     * @param {?} rowIdentity
     * @param {?} rowStateOrCellIndex
     * @param {?=} cellState
     * @return {?}
     */
    updateState(rowIdentity, rowStateOrCellIndex, cellState) {
        /** @type {?} */
        const currentRowState = this.cache.get(rowIdentity);
        if (currentRowState) {
            if (typeof rowStateOrCellIndex === 'number') {
                /** @type {?} */
                const currentCellState = currentRowState.cells[rowStateOrCellIndex];
                if (currentCellState) {
                    Object.assign(currentCellState, cellState);
                }
            }
            else {
                Object.assign(currentRowState, rowStateOrCellIndex);
            }
            /** @type {?} */
            const rowContext = this.findRowInView(rowIdentity);
            if (rowContext) {
                rowContext.fromState(currentRowState);
            }
        }
    }
    /**
     * Try to find a specific row, using the row identity, in the current view.
     * If the row is not in the view (or even not in the cache) it will return undefined, otherwise returns the row's context instance (`PblRowContext`)
     * @param {?} rowIdentity The row's identity. If a specific identity is used, please provide it otherwise provide the index of the row in the datasource.
     * @return {?}
     */
    findRowInView(rowIdentity) {
        /** @type {?} */
        const rowState = this.cache.get(rowIdentity);
        if (rowState) {
            /** @type {?} */
            const renderRowIndex = rowState.dataIndex - this.extApi.grid.ds.renderStart;
            /** @type {?} */
            const rowContext = this.viewCache.get(renderRowIndex);
            if (rowContext && rowContext.identity === rowIdentity) {
                return rowContext;
            }
        }
    }
    /**
     * @param {?} rowIdentity
     * @param {?=} offset
     * @param {?=} create
     * @return {?}
     */
    findRowInCache(rowIdentity, offset, create) {
        /** @type {?} */
        const rowState = this.cache.get(rowIdentity);
        if (!offset) {
            return rowState;
        }
        else {
            /** @type {?} */
            const dataIndex = rowState.dataIndex + offset;
            /** @type {?} */
            const identity = this.getRowIdentity(dataIndex);
            if (identity !== null) {
                /** @type {?} */
                let result = this.findRowInCache(identity);
                if (!result && create && dataIndex < this.extApi.grid.ds.length) {
                    result = PblRowContext.defaultState(identity, dataIndex, this.columnApi.columns.length);
                    this.cache.set(identity, result);
                }
                return result;
            }
        }
    }
    /**
     * @param {?} dataIndex
     * @param {?=} context
     * @return {?}
     */
    getRowIdentity(dataIndex, context) {
        const { ds } = this.extApi.grid;
        const { primary } = this.extApi.columnStore;
        /** @type {?} */
        const row = context ? context.$implicit : ds.source[dataIndex];
        if (!row) {
            return null;
        }
        else {
            return primary ? primary.getValue(row) : dataIndex;
        }
    }
    /**
     * @private
     * @param {?} index
     * @return {?}
     */
    findViewRef(index) {
        return (/** @type {?} */ (this.vcRef.get(index)));
    }
    /**
     * Find/Update/Create the `RowContext` for the provided `EmbeddedViewRef` at the provided render position.
     *
     * A `RowContext` object is a wrapper for the internal context of a row in `CdkTable` with the purpose of
     * extending it for the grid features.
     *
     * The process has 2 layers of cache:
     *
     * - `RowContext` objects are stored in a view cache which is synced with the `CdkTable` row outlet viewRefs.
     * Each view ref (row) has a matching record in the `RowContext` view cache.
     *
     * - `RowContextState` object are stored in a cache which is synced with the items in the data source.
     * Each item in the datasource has a matching row `RowContextState` item (lazy), which is used to persist context
     * when `RowContext` goes in/out of the viewport.
     *
     * @private
     * @param {?} viewRef The `EmbeddedViewRef` holding the context that the returned `RowContext` should wrap
     * @param {?} renderRowIndex The position of the view, relative to other rows.
     * The position is required for caching the context state when a specific row is thrown out of the viewport (virtual scroll).
     * Each `RowContext` gets a unique identity using the position relative to the current render range in the data source.
     * @return {?}
     */
    findRowContext(viewRef, renderRowIndex) {
        const { context } = viewRef;
        /** @type {?} */
        const dataIndex = this.extApi.grid.ds.renderStart + renderRowIndex;
        /** @type {?} */
        const identity = this.getRowIdentity(dataIndex, viewRef.context);
        /** @type {?} */
        let rowContext = (/** @type {?} */ (context.pblRowContext));
        if (!this.cache.has(identity)) {
            this.cache.set(identity, PblRowContext.defaultState(identity, dataIndex, this.columnApi.columns.length));
        }
        if (!rowContext) {
            rowContext = context.pblRowContext = new PblRowContext(identity, dataIndex, this.extApi);
            rowContext.updateContext(context);
            viewRef.onDestroy((/**
             * @return {?}
             */
            () => {
                this.viewCache.delete(renderRowIndex);
                context.pblRowContext = undefined;
            }));
        }
        else if (rowContext.identity !== identity) {
            // save old state before applying new state
            this.cache.set(rowContext.identity, rowContext.getState());
            rowContext.updateContext(context);
            // We
            /** @type {?} */
            const gap = dataIndex - rowContext.dataIndex;
            if (gap > 0) {
                /** @type {?} */
                const siblingViewRef = this.findViewRef(renderRowIndex + gap);
                /** @type {?} */
                const siblingRowContext = siblingViewRef && (/** @type {?} */ (siblingViewRef.context.pblRowContext));
                if (siblingRowContext) {
                    this.cache.set(siblingRowContext.identity, siblingRowContext.getState());
                }
            }
        }
        else {
            return rowContext;
        }
        rowContext.fromState(this.cache.get(identity));
        return rowContext;
    }
    /**
     * @private
     * @return {?}
     */
    getViewRect() {
        return this.extApi.grid.viewport.elementRef.nativeElement.getBoundingClientRect();
    }
    /**
     * @private
     * @param {?} curr
     * @return {?}
     */
    emitFocusChanged(curr) {
        this.focusChanged$.next({
            prev: this.focusChanged$.value.curr,
            curr,
        });
    }
    /**
     * @private
     * @return {?}
     */
    destroy() {
        this.focusChanged$.complete();
        this.selectionChanged$.complete();
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    ContextApi.prototype.viewCache;
    /**
     * @type {?}
     * @private
     */
    ContextApi.prototype.cache;
    /**
     * @type {?}
     * @private
     */
    ContextApi.prototype.vcRef;
    /**
     * @type {?}
     * @private
     */
    ContextApi.prototype.columnApi;
    /**
     * @type {?}
     * @private
     */
    ContextApi.prototype.activeFocused;
    /**
     * @type {?}
     * @private
     */
    ContextApi.prototype.activeSelected;
    /**
     * @type {?}
     * @private
     */
    ContextApi.prototype.focusChanged$;
    /**
     * @type {?}
     * @private
     */
    ContextApi.prototype.selectionChanged$;
    /**
     * Notify when the focus has changed.
     *
     * > Note that the notification is not immediate, it will occur on the closest micro-task after the change.
     * @type {?}
     */
    ContextApi.prototype.focusChanged;
    /**
     * Notify when the selected cells has changed.
     * @type {?}
     */
    ContextApi.prototype.selectionChanged;
    /**
     * @type {?}
     * @private
     */
    ContextApi.prototype.extApi;
}
/**
 * @param {?} viewRef
 * @param {?} viewPortRect
 * @param {?=} location
 * @return {?}
 */
function processOutOfView(viewRef, viewPortRect, location) {
    /** @type {?} */
    const el = viewRef.rootNodes[0];
    /** @type {?} */
    const rowContext = viewRef.context.pblRowContext;
    /** @type {?} */
    const elRect = el.getBoundingClientRect();
    /** @type {?} */
    let isInsideOfView;
    switch (location) {
        case 'top':
            isInsideOfView = elRect.bottom >= viewPortRect.top;
            break;
        case 'bottom':
            isInsideOfView = elRect.top <= viewPortRect.bottom;
            break;
        default:
            isInsideOfView = (elRect.bottom >= viewPortRect.top && elRect.top <= viewPortRect.bottom);
            break;
    }
    if (isInsideOfView) {
        if (!rowContext.outOfView) {
            return false;
        }
        rowContext.outOfView = false;
    }
    else {
        rowContext.outOfView = true;
    }
    return true;
}
